publisher/publisher.go

```diff
@@ -0,0 +1,274 @@
+			time.Sleep(backoff)
+			continue
+		} else {
+			if resp.StatusCode == http.StatusRequestTimeout || resp.StatusCode == http.StatusServiceUnavailable {
+				// Retry the request after either 10 seconds or the period specified
+				// by the Retry-After header
+				backoff = pub.submissionBackoff
+				if seconds, err := strconv.Atoi(resp.Header.Get("Retry-After")); err == nil {
+					backoff = time.Second * time.Duration(seconds)
+				} else {
+					fmt.Println(err)
+				}
+				time.Sleep(backoff)
+				continue
+			} else if resp.StatusCode != http.StatusOK {
+				// Not something we expect to happen, set error, break loop and log
+				// the error
+				// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+				pub.log.AuditErr(fmt.Errorf("Unexpected status code returned from CT log submission endpoint [%s]: Unexpected status code [%d]", log.URI, resp.StatusCode))
+				break
+			}
+			done = true
+			break
+		}
+	}
+
+	if !done {
+		pub.log.Warning(fmt.Sprintf(
+			"Unable to submit certificate to CT log [Serial: %s, Log URI: %s, Retries: %d]",
+			serial,
+			log.URI,
+			retries,
+		))
+		return fmt.Errorf("Unable to submit certificate")
+	}
+
+	if err := sct.CheckSignature(); err != nil {
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		pub.log.AuditErr(err)
+		return err
+	}
+
+	pub.log.Info(fmt.Sprintf(
+		"Submitted certificate to CT log [Serial: %s, Log URI: %s, Retries: %d, Signature: %x]",
+		serial,
+		log.URI,
+		retries, sct.Signature,
+	))
+
+	// Set certificate serial and add SCT to DB
+	sct.CertificateSerial = serial
+	err := pub.SA.AddSCTReceipt(sct)
+	if err != nil {
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		pub.log.AuditErr(fmt.Errorf(
+			"Error adding SCT receipt for [%s to %s]: %s",
+			sct.CertificateSerial,
+			log.URI,
+			err,
+		))
+		return err
+	}
+	pub.log.Info(fmt.Sprintf(
+		"Stored SCT receipt from CT log submission [Serial: %s, Log URI: %s]",
+		serial,
+		log.URI,
+	))
+	return nil
+}
+
+// SubmitToCT will submit the certificate represented by certDER to any CT
+// logs configured in pub.CT.Logs
+func (pub *PublisherImpl) SubmitToCT(der []byte) error {
+	cert, err := x509.ParseCertificate(der)
+	if err != nil {
+		pub.log.Err(fmt.Sprintf("Unable to parse certificate, %s", err))
+		return err
+	}
+
+	submission := ctSubmissionRequest{Chain: []string{base64.StdEncoding.EncodeToString(cert.Raw)}}
+	// Add all intermediate certificates needed for submission
+	submission.Chain = append(submission.Chain, pub.issuerBundle...)
+	jsonSubmission, err := json.Marshal(submission)
+	if err != nil {
+		pub.log.Err(fmt.Sprintf("Unable to marshal CT submission, %s", err))
+		return err
+	}
+
+	for _, ctLog := range pub.ctLogs {
+		err = pub.submitToCTLog(core.SerialToString(cert.SerialNumber), jsonSubmission, ctLog)
+		if err != nil {
+			pub.log.Err(err.Error())
+			continue
+		}
+	}
+
+	return err
```

Comment:
The returned error is currently only used by the Test... functions to confirm submission fails after retrying the max number of times. I think using the log mocking this could be replaced though.